### PR TITLE
remove subscriptions race

### PR DESCRIPTION
**Pull request by TyOverby**
Fix Issue # .
<Short description of the fix.>

- [x] I ran Unit Tests locally.

For significant contributions please make sure you have completed the following items:

- [ ] Changes in public surface reviewed
- [ ] Design discussion issue #
- [ ] CHANGELOG.md updated with one line description of the fix, and a link to the original issue.
- [ ] The PR will trigger build, unit tests, and functional tests automatically. If your PR was submitted from fork - mention one of committers to initiate the build for you.
	  If you want to to re-run the build/tests, the easiest way is to simply Close and Re-Open this same PR. (Just click 'close pull request' followed by 'open pull request' buttons at the bottom of the PR)

- Please follow [these] (https://github.com/Microsoft/ApplicationInsights-aspnetcore/blob/develop/Readme.md) instructions to build and test locally.

### CODE REVIEW

File: src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs [hunk 1]
```diff
@@ -48,11 +49,20 @@ public void Start()
...
         void IObserver<DiagnosticListener>.OnNext(DiagnosticListener value)
         {
-            foreach (var applicationInsightDiagnosticListener in this.diagnosticListeners)
+            lock (this.subscriptions)
             {
-                if (applicationInsightDiagnosticListener.ListenerName == value.Name)
+                if (shuttingDown)
                 {
-                    this.subscriptions.Add(value.SubscribeWithAdapter(applicationInsightDiagnosticListener));
+                    value.Dispose();
```
Comment (TyOverby) [inline src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs]: Q: Should this Dispose call be here?  

If we are the "owners" of `value` then I think it should be disposed, but is it possible that this value is shared?

Comment (TyOverby) [reply src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs]: I'll fix this, thanks for answering!

File: src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs [hunk 2]
```diff
@@ -48,11 +49,20 @@ public void Start()
         /// <inheritdoc />
         void IObserver<DiagnosticListener>.OnNext(DiagnosticListener value)
         {
-            foreach (var applicationInsightDiagnosticListener in this.diagnosticListeners)
+            lock (this.subscriptions)
```
[the remark this answers is not in the document]
Comment (TyOverby) [reply src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs]: Sounds good

File: src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs [hunk 3]
```diff
@@ -15,6 +15,7 @@ internal class ApplicationInsightsInitializer : IObserver<DiagnosticListener>, I
     {
         private readonly List<IDisposable> subscriptions;
         private readonly IEnumerable<IApplicationInsightDiagnosticListener> diagnosticListeners;
+        private bool shuttingDown = false;
```
[the remark this answers is not in the document]
Comment (TyOverby) [reply src/Microsoft.ApplicationInsights.AspNetCore/Implementation/ApplicationInsightsInitializer.cs]: Done.

**Comment by TyOverby** [comment]
@lmolkova I've made the request changes, however, I still believe that a lock is necessary.

Imagine the scenario: 

* <Thread 1> Disposal is called on the object
* <Thread 1> shuttingDown is set to `true`
* <Thread 2> OnNext is called on the object
* <Thread 2> shuttingDown is read as `false` (the above write happens concurrently with this read)
* <Thread 1> Subscriptions is looped over and its elements disposed
* <Thread 2> A new subscription is added to Subscriptions

With this model we mitigate the iterator invalidation exception (by making it use ConcurrentBag), but we run the risk of not disposing some elements in the bag (if those elements are added while Disposal is happening.

**Comment by pharring** [comment]
Suggestion:
Don't use a separate 'bool shuttingDown'.

Instead, remove `readonly` from `subscriptions` and, in `Dispose(bool disposing)` do this:
```C#
if (!disposing)
{
    return;
}

var subs = Interlocked.Exchange(ref this.subscriptions, null);
if (subs != null)
{
    foreach (var sub in subs) { sub.Dispose(); }
}
```

And, similarly, in `OnNext`:
```C#
var sub = Volatile.Read(ref this.subscirptions); // Could also do Interlocked.CompareExchange(ref this.subscriptions, null, null) to 'read' the current value
if (sub == null)
{
    // Shutting down
    return;
}
```

**Comment by TyOverby** [comment]
@cijothomas: done!

**Pull request merged**